# Info games enrichment

This script allows to enrich the `games_datasets` with games info from SteamDB

## PreProcessing

In [1]:
#!pip install thefuzz
#!pip install python-Levenshtein

In [2]:
import pandas as pd
import json
import collections
import re
import numpy as np
import requests
from thefuzz import fuzz
from thefuzz import process

C:\ProgramData\Anaconda3\lib\site-packages\thefuzz\fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


### Import Twitch games

In [3]:
games_tw = pd.read_csv('Output_datasets/Games_dataset_name.csv', index_col='ID_game')
games_tw

game_name
ID_game                                     
0        Monster Hunter Generations Ultimate
1                     Alice: Madness Returns
2                                  Hold Fast
3           Pokémon Trading Card Game Online
4                                   Witch It
...                                      ...
1139       Rise of Nations: Extended Edition
1140           Paper Mario: The Origami King
1141        The Inheritance of Crimson Manor
1142                                  Occult
1143                         Sign of Silence

[1144 rows x 1 columns]

#### Clean strings

Create dictionary with:

    {'ID_game' : { \
         'twitch_name': ' ', \
         'twitch_clean': ' ', \
         'steam_name' : ' ', \
         'steam_clean' : ' ', \
         'match_score': ' '\
         }}
         
where `twitch_clean` and `steam_clean` are:

    -strip
    -lower
    -alphanumeric + èòàìùüé
    -remove multiple spaces
    -remove all duplicates

In [4]:
match_dict = {}
clean = ''

for i in range(len(games_tw)):
    match_dict[i] = {}
    match_dict[i]['twitch_name'] = games_tw.game_name[i]
    clean = re.sub(r"[^a-zA-Z0-9èòàìùüé ']",'',games_tw.game_name[i]).lower().strip()
    clean = re.sub('\\s+', ' ', clean)
    match_dict[i]['twitch_clean'] = clean
    match_dict[i]['steam_name'] = ''
    match_dict[i]['steam_clean'] = ''
    match_dict[i]['match_score'] = ''
len(match_dict)

1144

In [5]:
match_df = pd.DataFrame.from_dict(match_dict,orient='index')
match_df

twitch_name  \
0     Monster Hunter Generations Ultimate   
1                  Alice: Madness Returns   
2                               Hold Fast   
3        Pokémon Trading Card Game Online   
4                                Witch It   
...                                   ...   
1139    Rise of Nations: Extended Edition   
1140        Paper Mario: The Origami King   
1141     The Inheritance of Crimson Manor   
1142                               Occult   
1143                      Sign of Silence   

                             twitch_clean steam_name steam_clean match_score  
0     monster hunter generations ultimate                                     
1                   alice madness returns                                     
2                               hold fast                                     
3        pokémon trading card game online                                     
4                                witch it                                     
...                                   ...        ...         ...         ...  
1139     rise of nations extended edition                                     
1140         paper mario the origami king                                     
1141     the inheritance of crimson manor                                     
1142                               occult                                     
1143                      sign of silence                                     

[1144 rows x 5 columns]

### Import Steam games

Reading the data from the file

In [6]:
with open('../DataCollection/Steam_games/steam_games.json', 'r') as f:
    games = json.load(f)

In [7]:
games['1599340'].keys()

dict_keys(['title', 'Developer', 'Publisher', 'Release Date', 'Last Update', '24h Player Peak', 'Followers', 'categories'])

#### Remove duplicates

In [8]:
steam_games = []

for keys in games:
    steam_games.append(games[keys]['title'])
print(len(steam_games))
print(len(set(steam_games)))

3498
3488


In [9]:
dupl_games =[item for item, count in collections.Counter(list(steam_games)).items() if count > 1]
dupl_games

["Conqueror's Blade",
 'DEAD OR ALIVE Xtreme Venus Vacation',
 'Fallout: New Vegas',
 'Ironsight',
 'Romance of the Three Kingdoms Hadou',
 'Wolfenstein II: The New Colossus',
 'Dishonored',
 'Wolfenstein: The New Order',
 'Anomaly Zone',
 'GRID']

In [10]:
keys_dupl = []
for elem in dupl_games:
    for key in games:
        if elem == games[key]['title']:
            keys_dupl.append(key)

Elimino i 10 duplicati presenti nella lista steam_games: dovrò poi risalire alle informazioni del json per capire in caso di match quali informazioni associare al gioco

In [11]:
steam_games = list(set(steam_games))

#### Clean strings

steam_clean ottenuta con:

    -strip
    -lower
    -alphanumeric + èòàìùüé
    -remove multiple spaces
    -remove all duplicates

In [12]:
steam_dict = {}

for i in range(len(steam_games)):
    steam_dict[i] = {}
    steam_dict[i]['steam_name'] = steam_games[i]
    clean = re.sub(r"[^a-zA-Z0-9èòàìùüé ']",'',steam_games[i]).lower().strip()
    clean = re.sub('\\s+', ' ', clean)
    steam_dict[i]['steam_clean'] = clean
len(steam_dict)

3488

In [13]:
steam_df = pd.DataFrame.from_dict(steam_dict,orient='index')
steam_df.head()

steam_name                     steam_clean
0               EA SPORTS™ FIFA 21               ea sports fifa 21
1                       RoBoRumble                      roborumble
2                        Anomaly 2                       anomaly 2
3  Stealth Inc 2: A Game of Clones  stealth inc 2 a game of clones
4                         Banished                        banished

In [14]:
steam_df[steam_df.steam_clean == '']

steam_name steam_clean
37            觅长生            
131        小黑盒加速器            
419          大千世界            
750         魔物讨伐团            
1351          星云纪            
1468    太荒初境（试玩版）            
1962        龙魂大冒险            
2140           春秋            
2643           真探            
2657  英雄黄昏-文字版三国志            
2785   軒轅劍參外傳 天之痕            
2787      仙剑奇侠传九野            
3046       飞火动态壁纸            
3424         枝江往事

Rimuovo queste 14 righe

In [15]:
steam_df = steam_df[steam_df.steam_clean != '']
len(steam_df)

3474

## Matching

### Perfect match between strings

This obtain a score of 100

In [16]:
len(list(set(match_df.twitch_name) & set(steam_df.steam_name)))

318

Aggiorno il dataframe inserendo i 318 match perfetti

In [17]:
for i in range(len(match_df)):
    if match_df.twitch_name[i] in list(steam_df.steam_name):
        match_df.match_score[i] = 100
        match_df.steam_name[i] = match_df.twitch_name[i]
match_df

twitch_name  \
0     Monster Hunter Generations Ultimate   
1                  Alice: Madness Returns   
2                               Hold Fast   
3        Pokémon Trading Card Game Online   
4                                Witch It   
...                                   ...   
1139    Rise of Nations: Extended Edition   
1140        Paper Mario: The Origami King   
1141     The Inheritance of Crimson Manor   
1142                               Occult   
1143                      Sign of Silence   

                             twitch_clean                         steam_name  \
0     monster hunter generations ultimate                                      
1                   alice madness returns             Alice: Madness Returns   
2                               hold fast                                      
3        pokémon trading card game online                                      
4                                witch it                           Witch It   
...                                   ...                                ...   
1139     rise of nations extended edition  Rise of Nations: Extended Edition   
1140         paper mario the origami king                                      
1141     the inheritance of crimson manor                                      
1142                               occult                                      
1143                      sign of silence                    Sign of Silence   

     steam_clean match_score  
0                             
1                        100  
2                             
3                             
4                        100  
...          ...         ...  
1139                     100  
1140                          
1141                          
1142                          
1143                     100  

[1144 rows x 5 columns]

### Perfect match between cleaned strings

Prima di effettuare il match devo assicurarmi di rimuovere i duplicati all'interno delle liste corrispondenti

In [18]:
import collections
[item for item, count in collections.Counter(list(match_df.twitch_clean)).items() if count > 1]

['f1 2021']

- 'F1 2021' e 'F1® 2021' \
in twitch_names con il cleaning si riducono alla stessa stringa
'f1 2021' \
quindi elimino duplicato.

In [19]:
match_df[match_df.twitch_clean=='f1 2021']

twitch_name twitch_clean steam_name steam_clean match_score
121     F1 2021      f1 2021                                   
482    F1® 2021      f1 2021   F1® 2021                     100

In [20]:
dupl = match_df[match_df.twitch_name == 'F1 2021']

Mantengo il duplicato che è già stato matchato

In [21]:
match_df = match_df.drop_duplicates(subset=['twitch_clean'], keep='last').reset_index(drop=True)

In [22]:
match_df[match_df.twitch_clean=='f1 2021']

twitch_name twitch_clean steam_name steam_clean match_score
481    F1® 2021      f1 2021   F1® 2021                     100

In [23]:
len(steam_df.steam_clean) - len(set(steam_df.steam_clean))

6

Ci sono 6 duplicati nella lista steam_clean: rimuovo righe associate dal dataframe

In [24]:
steam_df = steam_df.drop_duplicates(subset=["steam_clean"], keep='first').reset_index(drop=True)
steam_df

steam_name  \
0                             EA SPORTS™ FIFA 21   
1                                     RoBoRumble   
2                                      Anomaly 2   
3                Stealth Inc 2: A Game of Clones   
4                                       Banished   
...                                          ...   
3463                                     Tyranny   
3464      Viscera Cleanup Detail: Shadow Warrior   
3465  Truck Mechanic: Dangerous Paths - Prologue   
3466            Door Kickers 2: Task Force North   
3467                          Romero's Aftermath   

                                  steam_clean  
0                           ea sports fifa 21  
1                                  roborumble  
2                                   anomaly 2  
3              stealth inc 2 a game of clones  
4                                    banished  
...                                       ...  
3463                                  tyranny  
3464    viscera cleanup detail shadow warrior  
3465  truck mechanic dangerous paths prologue  
3466          door kickers 2 task force north  
3467                       romero's aftermath  

[3468 rows x 2 columns]

In [25]:
len(list(set(match_df.twitch_clean) & set(steam_df.steam_clean)))

402

Ci sono 84 nuovi match tra stringhe pulite: aggiorno il dataframe *match_df*

In [26]:
for i in range(len(match_df)):
    if (match_df.match_score[i] != 100) and (match_df.twitch_clean[i] in list(steam_df.steam_clean)):
        match_df.steam_clean[i] = match_df.twitch_clean[i]
        match_df.match_score[i] = 100

In [27]:
match_df.head(20)

twitch_name                         twitch_clean  \
0   Monster Hunter Generations Ultimate  monster hunter generations ultimate   
1                Alice: Madness Returns                alice madness returns   
2                             Hold Fast                            hold fast   
3      Pokémon Trading Card Game Online     pokémon trading card game online   
4                              Witch It                             witch it   
5                       Dante's Inferno                      dante's inferno   
6     The Stanley Parable: Ultra Deluxe     the stanley parable ultra deluxe   
7                        The Long Drive                       the long drive   
8                          Cyber Hunter                         cyber hunter   
9                           Rogue Lords                          rogue lords   
10      Assassin's Creed IV: Black Flag       assassin's creed iv black flag   
11                               INSIDE                               inside   
12          DRAGON BALL Z DOKKAN BATTLE          dragon ball z dokkan battle   
13                                 V696                                 v696   
14                                Hades                                hades   
15                        Stranded Deep                        stranded deep   
16             Legacy of Kain: Defiance              legacy of kain defiance   
17                               Rounds                               rounds   
18                            Drawful 2                            drawful 2   
19             Sensible World of Soccer             sensible world of soccer   

                           steam_name steam_clean match_score  
0                                                              
1              Alice: Madness Returns                     100  
2                                                              
3                                                              
4                            Witch It                     100  
5                                                              
6   The Stanley Parable: Ultra Deluxe                     100  
7                      The Long Drive                     100  
8                                                              
9                                                              
10                                                             
11                             INSIDE                     100  
12                                                             
13                                                             
14                              Hades                     100  
15                      Stranded Deep                     100  
16                                                             
17                                         rounds         100  
18                          Drawful 2                     100  
19

In [28]:
############ per risalire a steam_name
for i in range(len(match_df)):
    if (match_df.steam_clean[i] != ''):
        index = steam_df.loc[steam_df['steam_clean'] == match_df.steam_clean[i]].index[0]
        match_df.steam_name[i] = steam_df.steam_name[index]
match_df

twitch_name  \
0     Monster Hunter Generations Ultimate   
1                  Alice: Madness Returns   
2                               Hold Fast   
3        Pokémon Trading Card Game Online   
4                                Witch It   
...                                   ...   
1138    Rise of Nations: Extended Edition   
1139        Paper Mario: The Origami King   
1140     The Inheritance of Crimson Manor   
1141                               Occult   
1142                      Sign of Silence   

                             twitch_clean                         steam_name  \
0     monster hunter generations ultimate                                      
1                   alice madness returns             Alice: Madness Returns   
2                               hold fast                                      
3        pokémon trading card game online                                      
4                                witch it                           Witch It   
...                                   ...                                ...   
1138     rise of nations extended edition  Rise of Nations: Extended Edition   
1139         paper mario the origami king                                      
1140     the inheritance of crimson manor                                      
1141                               occult                                      
1142                      sign of silence                    Sign of Silence   

     steam_clean match_score  
0                             
1                        100  
2                             
3                             
4                        100  
...          ...         ...  
1138                     100  
1139                          
1140                          
1141                          
1142                     100  

[1143 rows x 5 columns]

In [29]:
match_100 = match_df[match_df.match_score==100]

In [30]:
len(match_df) - len(match_100)

741

### Fuzzy match between cleaned strings

with subsets matching starts of the strings

ancora 741 elementi da matchare ( uno non matcha)

In [31]:
match_df = match_df[match_df.match_score != 100].reset_index(drop=True)
len(match_df)

741

In [32]:
for elem in match_df.twitch_name:
    if len(elem) < 4:
        print(elem)

Saw
Art


In [33]:
for elem in steam_df.steam_clean:
    if elem.startswith('saw'):
        print(elem)
    if elem.startswith('art'):
        print(elem)   

artifact


Non ci sono match con 'Saw' e 'Art' di twitch.
Considero i subset di stringhe in steam_df.steam_clean con primi 4 elementi matchati

In [34]:
match_dict = match_df.to_dict('index')
match_dict

{0: {'twitch_name': 'Monster Hunter Generations Ultimate',
  'twitch_clean': 'monster hunter generations ultimate',
  'steam_name': '',
  'steam_clean': '',
  'match_score': ''},
 1: {'twitch_name': 'Hold Fast',
  'twitch_clean': 'hold fast',
  'steam_name': '',
  'steam_clean': '',
  'match_score': ''},
 2: {'twitch_name': 'Pokémon Trading Card Game Online',
  'twitch_clean': 'pokémon trading card game online',
  'steam_name': '',
  'steam_clean': '',
  'match_score': ''},
 3: {'twitch_name': "Dante's Inferno",
  'twitch_clean': "dante's inferno",
  'steam_name': '',
  'steam_clean': '',
  'match_score': ''},
 4: {'twitch_name': 'Cyber Hunter',
  'twitch_clean': 'cyber hunter',
  'steam_name': '',
  'steam_clean': '',
  'match_score': ''},
 5: {'twitch_name': 'Rogue Lords',
  'twitch_clean': 'rogue lords',
  'steam_name': '',
  'steam_clean': '',
  'match_score': ''},
 6: {'twitch_name': "Assassin's Creed IV: Black Flag",
  'twitch_clean': "assassin's creed iv black flag",
  'steam_na

In [35]:
for i in range(len(match_dict)):
    start = match_dict[i]['twitch_clean'][0:4]
    subset = []
    for elem in steam_df.steam_clean:
        if elem.startswith(start):
            subset.append(elem)
    if subset != []:
        score = {}  #key:nome_gioco / value: fuzz.ratio_score
        for elem in subset:
            score[elem] = fuzz.ratio(match_dict[i]['twitch_clean'], elem)
            keys = [k for k, v in score.items() if v == max(score.values())]
            number = list(set([v for k, v in score.items() if v == max(score.values())]))
            match_dict[i]['match_score'] = number[0]
            match_dict[i]['steam_clean'] = keys[0]

In [36]:
nuovo = pd.DataFrame.from_dict(match_dict, orient='index')

############ per risalire a steam_name
for i in range(len(nuovo)):
    if (nuovo.steam_clean[i] != ''):
        index = steam_df.loc[steam_df['steam_clean'] == nuovo.steam_clean[i]].index[0]
        nuovo.steam_name[i] = steam_df.steam_name[index]

In [37]:
fuzzy_match = nuovo[(nuovo.match_score != '')].sort_values(by=['match_score'], ascending=False).reset_index(drop=True)
fuzzy_match

twitch_name  \
0    Monkey Island 2 Special Edition: LeChuck's Rev...   
1                      Assassin's Creed IV: Black Flag   
2                            BlazBlue: Central Fiction   
3                          Sid Meier's Civilization VI   
4                        Assassin's Creed: Brotherhood   
..                                                 ...   
468                     Xenoverse: Per Aspera Ad Astra   
469                           Holdfast: Nations At War   
470                              WWE SLAM: Card Trader   
471           Xenoblade Chronicles: Definitive Edition   
472  Pocket Monsters Diamond & Pearl: Pokémon wo Sa...   

                                          twitch_clean  \
0    monkey island 2 special edition lechuck's revenge   
1                       assassin's creed iv black flag   
2                             blazblue central fiction   
3                          sid meier's civilization vi   
4                         assassin's creed brotherhood   
..                                                 ...   
468                      xenoverse per aspera ad astra   
469                            holdfast nations at war   
470                               wwe slam card trader   
471            xenoblade chronicles definitive edition   
472  pocket monsters diamond pearl pokémon wo sagas...   

                                            steam_name  \
0    Monkey Island™ 2 Special Edition: LeChuck’s Re...   
1                     Assassin’s Creed® IV Black Flag™   
2                              BlazBlue Centralfiction   
3                          Sid Meier's Civilization® V   
4                        Assassin’s Creed® Brotherhood   
..                                                 ...   
468                                         Xenobox VR   
469                                   Hold your houses   
470                                           WWE 2K22   
471                                         Xenobox VR   
472                                     Pocket Academy   

                                          steam_clean match_score  
0    monkey island 2 special edition lechucks revenge          99  
1                       assassins creed iv black flag          98  
2                             blazblue centralfiction          98  
3                          sid meier's civilization v          98  
4                         assassins creed brotherhood          98  
..                                                ...         ...  
468                                        xenobox vr          31  
469                                  hold your houses          31  
470                                          wwe 2k22          29  
471                                        xenobox vr          29  
472                                    pocket academy          28  

[473 rows x 5 columns]

In [38]:
len(fuzzy_match)

473

In [39]:
no_match = nuovo[(nuovo.match_score == '')]
len(no_match)

268

### Quality assessment and results

- match_score 100 --> inserisco tutte le informazioni

Controllo qualità dei fuzzy con score tra 80 e 100

In [40]:
fuzzy_match[fuzzy_match.match_score.between(80,100)][['twitch_name','steam_name','match_score']]

twitch_name  \
0   Monkey Island 2 Special Edition: LeChuck's Rev...   
1                     Assassin's Creed IV: Black Flag   
2                           BlazBlue: Central Fiction   
3                         Sid Meier's Civilization VI   
4                       Assassin's Creed: Brotherhood   
..                                                ...   
73                                           Speedrun   
74                                             DOOM 3   
75          The Dark Pictures Anthology: Man of Medan   
76                                     Kingdom Hearts   
77                                     Virtual Casino   

                                           steam_name match_score  
0   Monkey Island™ 2 Special Edition: LeChuck’s Re...          99  
1                    Assassin’s Creed® IV Black Flag™          98  
2                             BlazBlue Centralfiction          98  
3                         Sid Meier's Civilization® V          98  
4                       Assassin’s Creed® Brotherhood          98  
..                                                ...         ...  
73                                       SpeedRunners          80  
74                                               DOOM          80  
75        The Dark Pictures Anthology: House of Ashes          80  
76                                   Kingdom Heroes 8          80  
77                                        VirtualCast          80  

[78 rows x 3 columns]

Controllando manualmente la qualità dei match con score tra 80 e 100 otteniamo il seguente subset con 11 *twitch/steam_name* relativi allo stesso gioco: modifichiamo score con 100 e concateniamo

In [41]:
nuovi_100 = fuzzy_match.loc[[0,1,2,3,4,7,23,25,28,61,71]]

Elimino le righe dal df fuzzy_match

In [42]:
fuzzy_match.drop([0,1,2,3,4,7,23,25,28,61,71], axis=0, inplace=True)

In [43]:
len(fuzzy_match)

462

In [44]:
nuovi_100['match_score'] = 100
nuovi_100

twitch_name  \
0   Monkey Island 2 Special Edition: LeChuck's Rev...   
1                     Assassin's Creed IV: Black Flag   
2                           BlazBlue: Central Fiction   
3                         Sid Meier's Civilization VI   
4                       Assassin's Creed: Brotherhood   
7                                    Puyo Puyo Tetris   
23                        Tom Clancy's The Division 2   
25                                           Infernax   
28                    Dark Messiah of Might and Magic   
61                 Friday the 13th: The Computer Game   
71                   The Binding of Isaac: Repentance   

                                         twitch_clean  \
0   monkey island 2 special edition lechuck's revenge   
1                      assassin's creed iv black flag   
2                            blazblue central fiction   
3                         sid meier's civilization vi   
4                        assassin's creed brotherhood   
7                                    puyo puyo tetris   
23                        tom clancy's the division 2   
25                                           infernax   
28                    dark messiah of might and magic   
61                  friday the 13th the computer game   
71                    the binding of isaac repentance   

                                           steam_name  \
0   Monkey Island™ 2 Special Edition: LeChuck’s Re...   
1                    Assassin’s Creed® IV Black Flag™   
2                             BlazBlue Centralfiction   
3                         Sid Meier's Civilization® V   
4                       Assassin’s Creed® Brotherhood   
7                                   Puyo Puyo™Tetris®   
23                         Tom Clancy’s The Division™   
25                                            Inferna   
28                      Dark Messiah of Might & Magic   
61                          Friday the 13th: The Game   
71                      The Binding of Isaac: Rebirth   

                                         steam_clean  match_score  
0   monkey island 2 special edition lechucks revenge          100  
1                      assassins creed iv black flag          100  
2                            blazblue centralfiction          100  
3                         sid meier's civilization v          100  
4                        assassins creed brotherhood          100  
7                                    puyo puyotetris          100  
23                          tom clancys the division          100  
25                                           inferna          100  
28                       dark messiah of might magic          100  
61                          friday the 13th the game          100  
71                      the binding of isaac rebirth          100

In [45]:
match_100 = pd.concat([match_100,nuovi_100]).reset_index(drop=True)

Sempre nei match con score tra 80 e 100 otteniamo il seguente subset con 8 twitch/steam_name relativi giochi diversi: eliminiamo lo score

In [46]:
nuovi_no_match = fuzzy_match.loc[[27,40,56,62,67,69,74,76,77]]

Elimino le righe dal df fuzzy_match

In [47]:
fuzzy_match.drop([27,40,56,62,67,69,74,76,77], axis=0, inplace=True)
len(fuzzy_match)

453

In [48]:
nuovi_no_match['match_score'] = ''
nuovi_no_match

twitch_name                  twitch_clean  \
27         Stronghold Crusader II        stronghold crusader ii   
40                  BUS SIMULATOR                 bus simulator   
56                        FIFA 21                       fifa 21   
62                Expedition Zero               expedition zero   
67  The King of Fighters All-Star  the king of fighters allstar   
69              Tomb Raider: Gold              tomb raider gold   
74                         DOOM 3                        doom 3   
76                 Kingdom Hearts                kingdom hearts   
77                 Virtual Casino                virtual casino   

                 steam_name              steam_clean match_score  
27    Stronghold Crusader 2    stronghold crusader 2              
40         Bus Simulator 16         bus simulator 16              
56                  FIFA 22                  fifa 22              
62        Expeditions: Rome         expeditions rome              
67  THE KING OF FIGHTERS XV  the king of fighters xv              
69              Tomb Raider              tomb raider              
74                     DOOM                     doom              
76         Kingdom Heroes 8         kingdom heroes 8              
77              VirtualCast              virtualcast

In [49]:
no_match = pd.concat([no_match,nuovi_no_match]).reset_index(drop=True)
len(no_match)

277

#### Results

In [50]:
len(match_100) + len(fuzzy_match) + len(no_match) + len(dupl)

1144

In [51]:
games_total = pd.concat([match_100,fuzzy_match,no_match,dupl]).reset_index(drop=True)
games_total

twitch_name                      twitch_clean  \
0                Alice: Madness Returns             alice madness returns   
1                              Witch It                          witch it   
2     The Stanley Parable: Ultra Deluxe  the stanley parable ultra deluxe   
3                        The Long Drive                    the long drive   
4                                INSIDE                            inside   
...                                 ...                               ...   
1139                  Tomb Raider: Gold                  tomb raider gold   
1140                             DOOM 3                            doom 3   
1141                     Kingdom Hearts                    kingdom hearts   
1142                     Virtual Casino                    virtual casino   
1143                            F1 2021                           f1 2021   

                             steam_name       steam_clean match_score  
0                Alice: Madness Returns                           100  
1                              Witch It                           100  
2     The Stanley Parable: Ultra Deluxe                           100  
3                        The Long Drive                           100  
4                                INSIDE                           100  
...                                 ...               ...         ...  
1139                        Tomb Raider       tomb raider              
1140                               DOOM              doom              
1141                   Kingdom Heroes 8  kingdom heroes 8              
1142                        VirtualCast       virtualcast              
1143                                                                   

[1144 rows x 5 columns]

In [52]:
#games_total.to_csv('games_total.csv')

Dall'analisi dei fuzzy_match con score maggiore di 70, estriamo le righe relative a match maggiori o uguali a score 79, perchè relative a match tra edizioni differenti del medesimo videogioco: per questo motivo questi match corrispondono ad un match di categorie di gioco

In [53]:
fuzzy_high = fuzzy_match[fuzzy_match.match_score.between(79,100)]
len(fuzzy_high)

66

## Merging

- match_100 : enrichment con 'title','Developer','Publisher','Release Date','categories';

- fuzzy_high : enrichment con 'title','Developer','Publisher','categories'

controllo i 10 giochi duplicati

In [54]:
for row in range(len(games_total)):
    if games_total.steam_name[row] in dupl_games:
        print(row)

45


In [55]:
dupl_games

["Conqueror's Blade",
 'DEAD OR ALIVE Xtreme Venus Vacation',
 'Fallout: New Vegas',
 'Ironsight',
 'Romance of the Three Kingdoms Hadou',
 'Wolfenstein II: The New Colossus',
 'Dishonored',
 'Wolfenstein: The New Order',
 'Anomaly Zone',
 'GRID']

In [56]:
for row in range(len(games_total)):
    if games_total.steam_name[row] in dupl_games:
        print(games_total.twitch_name[row])

Fallout: New Vegas


Tra i giochi per cui faremo enrichment, 'Fallout: New Vegas' presenta un duplicato in steam_db

In [57]:
for key in games:
    if games[key]['title'] == 'Fallout: New Vegas':
        print(key,games[key])

22380 {'title': 'Fallout: New Vegas', 'Developer': 'Obsidian Entertainment', 'Publisher': 'Bethesda Softworks', 'Release Date': '19 October 2010', '24h Player Peak': '5,049', 'Followers': '143,057', 'categories': ['Open World', 'RPG', 'Post-apocalyptic', 'Singleplayer', 'Moddable', 'Exploration']}
22490 {'title': 'Fallout: New Vegas', 'Developer': 'Obsidian Entertainment', 'Publisher': 'Bethesda Softworks', 'Release Date': '21 October 2010', '24h Player Peak': '817', 'Followers': '9,254', 'categories': ['RPG', 'Post-apocalyptic', 'Open World', 'Atmospheric', 'Story Rich', 'Action']}


Scelgo di arricchirre con le info del gioco associato alla data di rilascio '19 october 2010' [22380], con maggiori followers:
il successivo [22490], data di rilascio due giorni dopo, è probabilmente una patch.    

In [58]:
to_enrich = pd.concat([match_100,fuzzy_high])
to_enrich

twitch_name                      twitch_clean  \
0              Alice: Madness Returns             alice madness returns   
1                            Witch It                          witch it   
2   The Stanley Parable: Ultra Deluxe  the stanley parable ultra deluxe   
3                      The Long Drive                    the long drive   
4                              INSIDE                            inside   
..                                ...                               ...   
81                      Dragon Ball Z                     dragon ball z   
82          Assassin's Creed Valhalla         assassin's creed valhalla   
83          Dissidia Final Fantasy NT         dissidia final fantasy nt   
84               Call of Duty: Mobile               call of duty mobile   
85              Batman: Arkham Asylum              batman arkham asylum   

                                steam_name  \
0                   Alice: Madness Returns   
1                                 Witch It   
2        The Stanley Parable: Ultra Deluxe   
3                           The Long Drive   
4                                   INSIDE   
..                                     ...   
81                    DRAGON BALL FighterZ   
82                      Assassin's Creed 2   
83  DISSIDIA FINAL FANTASY NT Free Edition   
84                         Call of Duty® 2   
85                     Batman: Arkham City   

                               steam_clean match_score  
0                                                  100  
1                                                  100  
2                                                  100  
3                                                  100  
4                                                  100  
..                                     ...         ...  
81                    dragon ball fighterz          79  
82                      assassin's creed 2          79  
83  dissidia final fantasy nt free edition          79  
84                          call of duty 2          79  
85                      batman arkham city          79  

[479 rows x 5 columns]

In [59]:
games_tw = games_tw.rename(columns={'game_name':'twitch_name'})
games_tw

twitch_name
ID_game                                     
0        Monster Hunter Generations Ultimate
1                     Alice: Madness Returns
2                                  Hold Fast
3           Pokémon Trading Card Game Online
4                                   Witch It
...                                      ...
1139       Rise of Nations: Extended Edition
1140           Paper Mario: The Origami King
1141        The Inheritance of Crimson Manor
1142                                  Occult
1143                         Sign of Silence

[1144 rows x 1 columns]

In [60]:
info_df = games_tw.merge(to_enrich[['twitch_name','steam_name','match_score']] , on= 'twitch_name', how='left')
info_df['Developer'] = ''
info_df['Publisher'] = ''
info_df['Release_Date'] = ''
info_df['categories'] = ''
info_df

twitch_name                         steam_name  \
0     Monster Hunter Generations Ultimate                                NaN   
1                  Alice: Madness Returns             Alice: Madness Returns   
2                               Hold Fast                                NaN   
3        Pokémon Trading Card Game Online                                NaN   
4                                Witch It                           Witch It   
...                                   ...                                ...   
1139    Rise of Nations: Extended Edition  Rise of Nations: Extended Edition   
1140        Paper Mario: The Origami King                                NaN   
1141     The Inheritance of Crimson Manor                                NaN   
1142                               Occult                                NaN   
1143                      Sign of Silence                    Sign of Silence   

     match_score Developer Publisher Release_Date categories  
0            NaN                                              
1            100                                              
2            NaN                                              
3            NaN                                              
4            100                                              
...          ...       ...       ...          ...        ...  
1139         100                                              
1140         NaN                                              
1141         NaN                                              
1142         NaN                                              
1143         100                                              

[1144 rows x 7 columns]

In [61]:
games_df = pd.DataFrame.from_dict(games,orient='index')
games_df

title              Developer  \
1599340                          Lost Ark          Smilegate RPG   
730      Counter-Strike: Global Offensive                  Valve   
570                                Dota 2                  Valve   
578080                PUBG: BATTLEGROUNDS          KRAFTON, Inc.   
1172470                     Apex Legends™  Respawn Entertainment   
...                                   ...                    ...   
220820                   Zombie Driver HD           Exor Studios   
491030                     The Last Error  Anatoliy Loginovskikh   
370510        Marble Mayhem: Fragile Ball    Bimboosoft Co.,Ltd.   
524680              Escape From Nazi Labs    Manic Mind Game Lab   
686230                           OutSplit               Riviysky   

               Publisher      Release Date   Last Update 24h Player Peak  \
1599340     Amazon Games  11 February 2022   9 June 2022         880,727   
730                  NaN    21 August 2012   2 June 2022         840,048   
570                  NaN       9 July 2013  13 June 2022         672,736   
578080               NaN  21 December 2017   9 June 2022         367,516   
1172470  Electronic Arts   4 November 2020   7 June 2022         302,819   
...                  ...               ...           ...             ...   
220820               NaN   17 October 2012           NaN               9   
491030               NaN           Unknown           NaN               0   
370510    Senpai Studios       7 July 2015           NaN               1   
524680               NaN           Unknown           NaN               0   
686230               NaN           Unknown           NaN               0   

         Followers                                         categories  
1599340    268,276  [MMORPG, Free to Play, Action RPG, RPG, Action...  
730      3,155,868  [FPS, Shooter, Multiplayer, Competitive, Actio...  
570      1,981,324  [Free to Play, MOBA, Multiplayer, Strategy, eS...  
578080   3,032,121  [Survival, Shooter, Multiplayer, Battle Royale...  
1172470    323,408  [Free to Play, Battle Royale, Multiplayer, Sho...  
...            ...                                                ...  
220820       6,586  [Zombies, Racing, Top-Down, Gore, Violent, Ope...  
491030         108                      [Action, 2D, Puzzle, Shooter]  
370510         696     [Casual, Indie, Puzzle, Difficult, Arcade, 2D]  
524680         325  [Action, RPG, Indie, Perma Death, Retro, Proce...  
686230       1,139                                    [Indie, Puzzle]  

[3498 rows x 8 columns]

Aggiungo le informazioni di 'Fallout: New Vegas' relative alla chiave '22380', escludendo la patch successiva

In [62]:
for i in range(len(info_df)):
    if info_df.twitch_name[i] == 'Fallout: New Vegas': #22380
        info_df.Developer[i] = games_df['Developer']['22380']
        info_df.Publisher[i] = games_df['Publisher']['22380']
        info_df.Release_Date[i] = games_df['Release Date']['22380']
        info_df.Developer[i] = games_df['Developer']['22380']
        info_df.categories[i] = games_df['categories']['22380']

In [63]:
info_df.columns

Index(['twitch_name', 'steam_name', 'match_score', 'Developer', 'Publisher',
       'Release_Date', 'categories'],
      dtype='object')

Per match perfetti con score 100 inserisco tutte le informazioni presenti nel json

In [64]:
for i in range(len(info_df)):
    if (info_df.twitch_name[i] != 'Fallout: New Vegas') & (info_df.match_score[i] == 100):
        index = games_df.loc[games_df['title'] == info_df.steam_name[i]].index[0]
        info_df.Developer[i] = games_df['Developer'][index]
        info_df.Publisher[i] = games_df['Publisher'][index]
        info_df.Release_Date[i] = games_df['Release Date'][index]
        info_df.categories[i] = games_df['categories'][index]

In [65]:
info_df['match_score'] = info_df['match_score'].fillna(0)

Per match con score tra 79 e 99 inserisco le informazioni 'developer','publisher','categories' presenti nel json

In [66]:
for i in range(len(info_df)):
    if (info_df.twitch_name[i] != 'Fallout: New Vegas') & (79 <= info_df['match_score'][i] <= 99):
        index = games_df.loc[games_df['title'] == info_df.steam_name[i]].index[0]
        info_df.Developer[i] = games_df['Developer'][index]
        info_df.Publisher[i] = games_df['Publisher'][index]
        info_df.categories[i] = games_df['categories'][index]

C:\Users\gianl\AppData\Local\Temp/ipykernel_12756/779521762.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  info_df.Developer[i] = games_df['Developer'][index]
C:\Users\gianl\AppData\Local\Temp/ipykernel_12756/779521762.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  info_df.Publisher[i] = games_df['Publisher'][index]
C:\Users\gianl\AppData\Local\Temp/ipykernel_12756/779521762.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
 

In [67]:
info_df['Release_Date'] = info_df['Release_Date'].replace(['Unknown', ''], np.nan, regex=True)
info_df

twitch_name                         steam_name  \
0     Monster Hunter Generations Ultimate                                NaN   
1                  Alice: Madness Returns             Alice: Madness Returns   
2                               Hold Fast                                NaN   
3        Pokémon Trading Card Game Online                                NaN   
4                                Witch It                           Witch It   
...                                   ...                                ...   
1139    Rise of Nations: Extended Edition  Rise of Nations: Extended Edition   
1140        Paper Mario: The Origami King                                NaN   
1141     The Inheritance of Crimson Manor                                NaN   
1142                               Occult                                NaN   
1143                      Sign of Silence                    Sign of Silence   

      match_score          Developer               Publisher     Release_Date  \
0               0                                                         NaN   
1             100  Spicy Horse Games         Electronic Arts     17 June 2011   
2               0                                                         NaN   
3               0                                                         NaN   
4             100  Barrel Roll Games  Daedalic Entertainment  22 October 2020   
...           ...                ...                     ...              ...   
1139          100        SkyBox Labs       Xbox Game Studios     12 June 2014   
1140            0                                                         NaN   
1141            0                                                         NaN   
1142            0                                                         NaN   
1143          100          Renderise                     NaN   1 October 2020   

                                             categories  
0                                                        
1     [3D Platformer, Action-Adventure, Hack and Sla...  
2                                                        
3                                                        
4     [Multiplayer, Funny, Casual, Action, Indie, Hi...  
...                                                 ...  
1139  [Strategy, RTS, Classic, Multiplayer, Historic...  
1140                                                     
1141                                                     
1142                                                     
1143  [Horror, Co-op, Multiplayer, Atmospheric, Sing...  

[1144 rows x 7 columns]

In [68]:
#info_df = info_df.replace(np.nan, '', regex=True)
info_df.drop(['match_score'], axis = 1, inplace = True)

In [69]:
info_df['Release_Date'] = pd.to_datetime(info_df['Release_Date'], format='%d %B %Y')
info_df

twitch_name                         steam_name  \
0     Monster Hunter Generations Ultimate                                NaN   
1                  Alice: Madness Returns             Alice: Madness Returns   
2                               Hold Fast                                NaN   
3        Pokémon Trading Card Game Online                                NaN   
4                                Witch It                           Witch It   
...                                   ...                                ...   
1139    Rise of Nations: Extended Edition  Rise of Nations: Extended Edition   
1140        Paper Mario: The Origami King                                NaN   
1141     The Inheritance of Crimson Manor                                NaN   
1142                               Occult                                NaN   
1143                      Sign of Silence                    Sign of Silence   

              Developer               Publisher Release_Date  \
0                                                        NaT   
1     Spicy Horse Games         Electronic Arts   2011-06-17   
2                                                        NaT   
3                                                        NaT   
4     Barrel Roll Games  Daedalic Entertainment   2020-10-22   
...                 ...                     ...          ...   
1139        SkyBox Labs       Xbox Game Studios   2014-06-12   
1140                                                     NaT   
1141                                                     NaT   
1142                                                     NaT   
1143          Renderise                     NaN   2020-10-01   

                                             categories  
0                                                        
1     [3D Platformer, Action-Adventure, Hack and Sla...  
2                                                        
3                                                        
4     [Multiplayer, Funny, Casual, Action, Indie, Hi...  
...                                                 ...  
1139  [Strategy, RTS, Classic, Multiplayer, Historic...  
1140                                                     
1141                                                     
1142                                                     
1143  [Horror, Co-op, Multiplayer, Atmospheric, Sing...  

[1144 rows x 6 columns]

In [70]:
info_df.index.name = 'idGame'
info_df.columns = ['gameName', 'steamName', 'developer', 'publisher', 'releaseDate', 'categories']
info_df

gameName  \
idGame                                        
0       Monster Hunter Generations Ultimate   
1                    Alice: Madness Returns   
2                                 Hold Fast   
3          Pokémon Trading Card Game Online   
4                                  Witch It   
...                                     ...   
1139      Rise of Nations: Extended Edition   
1140          Paper Mario: The Origami King   
1141       The Inheritance of Crimson Manor   
1142                                 Occult   
1143                        Sign of Silence   

                                steamName          developer  \
idGame                                                         
0                                     NaN                      
1                  Alice: Madness Returns  Spicy Horse Games   
2                                     NaN                      
3                                     NaN                      
4                                Witch It  Barrel Roll Games   
...                                   ...                ...   
1139    Rise of Nations: Extended Edition        SkyBox Labs   
1140                                  NaN                      
1141                                  NaN                      
1142                                  NaN                      
1143                      Sign of Silence          Renderise   

                     publisher releaseDate  \
idGame                                       
0                                      NaT   
1              Electronic Arts  2011-06-17   
2                                      NaT   
3                                      NaT   
4       Daedalic Entertainment  2020-10-22   
...                        ...         ...   
1139         Xbox Game Studios  2014-06-12   
1140                                   NaT   
1141                                   NaT   
1142                                   NaT   
1143                       NaN  2020-10-01   

                                               categories  
idGame                                                     
0                                                          
1       [3D Platformer, Action-Adventure, Hack and Sla...  
2                                                          
3                                                          
4       [Multiplayer, Funny, Casual, Action, Indie, Hi...  
...                                                   ...  
1139    [Strategy, RTS, Classic, Multiplayer, Historic...  
1140                                                       
1141                                                       
1142                                                       
1143    [Horror, Co-op, Multiplayer, Atmospheric, Sing...  

[1144 rows x 6 columns]

In [71]:
info_df['categories']=info_df['categories'].astype(str)
info_df['categories']=info_df['categories'].replace(["\[","\]","'"], "",regex=True)
info_df['categories']=info_df['categories'].replace(',', ';',regex=True)
info_df

gameName  \
idGame                                        
0       Monster Hunter Generations Ultimate   
1                    Alice: Madness Returns   
2                                 Hold Fast   
3          Pokémon Trading Card Game Online   
4                                  Witch It   
...                                     ...   
1139      Rise of Nations: Extended Edition   
1140          Paper Mario: The Origami King   
1141       The Inheritance of Crimson Manor   
1142                                 Occult   
1143                        Sign of Silence   

                                steamName          developer  \
idGame                                                         
0                                     NaN                      
1                  Alice: Madness Returns  Spicy Horse Games   
2                                     NaN                      
3                                     NaN                      
4                                Witch It  Barrel Roll Games   
...                                   ...                ...   
1139    Rise of Nations: Extended Edition        SkyBox Labs   
1140                                  NaN                      
1141                                  NaN                      
1142                                  NaN                      
1143                      Sign of Silence          Renderise   

                     publisher releaseDate  \
idGame                                       
0                                      NaT   
1              Electronic Arts  2011-06-17   
2                                      NaT   
3                                      NaT   
4       Daedalic Entertainment  2020-10-22   
...                        ...         ...   
1139         Xbox Game Studios  2014-06-12   
1140                                   NaT   
1141                                   NaT   
1142                                   NaT   
1143                       NaN  2020-10-01   

                                               categories  
idGame                                                     
0                                                          
1       3D Platformer; Action-Adventure; Hack and Slas...  
2                                                          
3                                                          
4       Multiplayer; Funny; Casual; Action; Indie; Hid...  
...                                                   ...  
1139    Strategy; RTS; Classic; Multiplayer; Historica...  
1140                                                       
1141                                                       
1142                                                       
1143    Horror; Co-op; Multiplayer; Atmospheric; Singl...  

[1144 rows x 6 columns]

In [72]:
info_df.to_csv('Output_datasets/games_dataset.csv')